lldb/source/Plugins/Language/Swift/SwiftUnsafeTypes.cpp

```diff
@@ -89,35 +103,104 @@ bool SwiftUnsafeBufferPointer::Update() {
     argument_type = SwiftASTContext::GetGenericArgumentType(type, 0);
 
-  if (!argument_type.IsValid())
-    return false;
-
-  m_elem_type = argument_type;
+  if (argument_type.IsValid())
+    m_elem_type = argument_type;
 
   ValueObjectSP pointer_value_sp = some_value_sp->GetChildAtIndex(0, true);
   if (!pointer_value_sp)
     return false;
 
-  addr_t addr = pointer_value_sp->GetValueAsUnsigned(LLDB_INVALID_ADDRESS);
+  return pointer_value_sp->GetValueAsUnsigned(LLDB_INVALID_ADDRESS);
+}
 
-  if (!addr || addr == LLDB_INVALID_ADDRESS)
+class SwiftUnsafeRawBufferPointer : public SwiftUnsafeBufferPointer {
+public:
+  SwiftUnsafeRawBufferPointer(ValueObject &valobj);
+  bool Update() override;
+
+private:
+  addr_t m_end_addr;
+};
+
+SwiftUnsafeRawBufferPointer::SwiftUnsafeRawBufferPointer(ValueObject &valobj)
+    : SwiftUnsafeBufferPointer(valobj) {}
+
+bool SwiftUnsafeRawBufferPointer::Update() {
+  if (!m_valobj.GetNumChildren())
     return false;
 
+  // Here is the layout of Swift's UnsafeRaw[Mutable]BufferPointer.
+  //
+  // ▿ UnsafeRawBufferPointer
+  //   ▿ _position : Optional<UnsafeRawPointer<Int>>
+  //     ▿ some : UnsafeRawPointer<Int>
+  //       - pointerValue : Int
+  //   ▿ _end : Optional<UnsafeRawPointer<Int>>
+  //     ▿ some : UnsafeRawPointer<Int>
+  //       - pointerValue : Int
+  //
+  // The structure has 2 Optional UnsafePointer to the buffer's start address
+  // and end address. To access the pointer address, lldb unfolds every value
+  // object child until reaching `pointerValue`.
+
+  addr_t addr = GetAddress("_position");
+  if (!addr || addr == LLDB_INVALID_ADDRESS)
+    return false;
   m_start_addr = addr;
 
+  addr = GetAddress("_end");
+  if (!addr || addr == LLDB_INVALID_ADDRESS)
+    return false;
+  m_end_addr = addr;
+
+  if (!m_elem_type.IsValid()) {
+    auto ast_context = m_valobj.GetScratchSwiftASTContext();
+    if (!ast_context)
+      return false;
+
+    swift::TypeBase *swift_type =
+        ast_context->ReconstructType(ConstString("$ss5UInt8VD")); // Swift.UInt8
+    if (!swift_type)
+      return false;
+
+    CompilerType compiler_type = ast_context->GetCompilerType(swift_type);
+    if (!compiler_type.IsValid())
+      return false;
+
+    m_elem_type = compiler_type;
+  }
+
+  auto opt_type_size = m_elem_type.GetByteSize(m_valobj.GetTargetSP().get());
+
+  if (!opt_type_size)
+    return false;
+
+  m_count = (m_end_addr - m_start_addr) / *opt_type_size;
+
   return true;
 }
 
 bool lldb_private::formatters::swift::UnsafeBufferPointerSummaryProvider(
     ValueObject &valobj, Stream &stream, const TypeSummaryOptions &options) {
 
-  SwiftUnsafeBufferPointer swift_ubp(valobj);
+  size_t count = 0;
+  addr_t addr = LLDB_INVALID_ADDRESS;
 
-  if (!swift_ubp.Update())
-    return false;
+  llvm::StringRef type_name(valobj.GetTypeName().GetCString());
 
-  size_t count = swift_ubp.GetCount();
-  addr_t addr = swift_ubp.GetStartAddress();
+  if (!type_name.contains("Raw")) {
```

Comment:
Be nice to unify this.  If you had a factory that returned the right kind of object as a SwiftUnsafeBufferPointerUP then you wouldn't need the if/else